lib/fizzy/instantiate.cpp

```diff
@@ -227,6 +227,62 @@ Value eval_constant_expression(ConstantExpression expr,
         return globals[global_idx - imported_globals.size()];
 }
 
+ExternalFunction find_imported_function(const std::string& module, const std::string& name,
+    FuncType module_func_type, const std::vector<ImportedFunction>& imported_functions)
+{
+    const auto it = std::find_if(imported_functions.begin(), imported_functions.end(),
+        [module, name](const auto& func) { return module == func.module && name == func.name; });
+
+    if (it == imported_functions.end())
+    {
+        throw instantiate_error{"imported function " + module + "." + name + " is required"};
+    }
+
+    if (module_func_type.inputs != it->inputs)
+    {
+        throw instantiate_error{"function " + module + "." + name +
+                                " input types don't match imported function in module"};
+    }
+    if (module_func_type.outputs.empty() && it->output.has_value())
+    {
+        throw instantiate_error{
+            "function " + module + "." + name + " has output but is defined void in module"};
+    }
+    if (!module_func_type.outputs.empty() &&
+        (!it->output.has_value() || module_func_type.outputs[0] != *it->output))
+    {
+        throw instantiate_error{"function " + module + "." + name +
+                                " output type doesn't match imported function in module"};
+    }
+
+    return {it->function, module_func_type};
+}
+
+ExternalGlobal find_imported_global(const std::string& module, const std::string& name,
+    GlobalType module_global_type, const std::vector<ImportedGlobal>& imported_globals)
+{
+    const auto it = std::find_if(imported_globals.begin(), imported_globals.end(),
+        [module, name](const auto& func) { return module == func.module && name == func.name; });
+
+    if (it == imported_globals.end())
+    {
+        throw instantiate_error{"imported global " + module + "." + name + " is required"};
+    }
+
+    if (module_global_type.value_type != it->type)
+    {
+        throw instantiate_error{"global " + module + "." + name +
+                                " value type doesn't match imported global in module"};
+    }
+    if (module_global_type.is_mutable != it->is_mutable)
+    {
+        throw instantiate_error{"global " + module + "." + name +
+                                " mutability doesn't match imported global in module"};
+    }
+
```

Comment:
Should we check here also that `it->value != nullptr` ?
(note that it's checked in `instantiate` anyway)

Reply:
I think it is fine as is. Perhaps a comment here could be useful to say that `instantiate` will validate the presence of a value.

Reply:
Added a comment.